kernel/src/log_segment/mod.rs

```diff
@@ -0,0 +1,374 @@
+        // filter out log files that do not contain metadata or protocol information
+        static META_PREDICATE: LazyLock<Option<ExpressionRef>> = LazyLock::new(|| {
+            Some(Arc::new(Expression::or(
+                Expression::column([METADATA_NAME, "id"]).is_not_null(),
+                Expression::column([PROTOCOL_NAME, "minReaderVersion"]).is_not_null(),
+            )))
+        });
+        // read the same protocol and metadata schema for both commits and checkpoints
+        self.replay(engine, schema.clone(), schema, META_PREDICATE.clone())
+    }
+}
+
+/// Returns a fallible iterator of [`ParsedLogPath`] that are between the provided `start_version` (inclusive)
+/// and `end_version` (inclusive). [`ParsedLogPath`] may be a commit or a checkpoint.  If `start_version` is
+/// not specified, the files will begin from version number 0. If `end_version` is not specified, files up to
+/// the most recent version will be included.
+///
+/// Note: this calls [`FileSystemClient::list_from`] to get the list of log files.
+fn list_log_files(
+    fs_client: &dyn FileSystemClient,
+    log_root: &Url,
+    start_version: impl Into<Option<Version>>,
+    end_version: impl Into<Option<Version>>,
+) -> DeltaResult<impl Iterator<Item = DeltaResult<ParsedLogPath>>> {
+    let start_version = start_version.into().unwrap_or(0);
+    let end_version = end_version.into();
+
+    let version_prefix = format!("{:020}", start_version);
+    let start_from = log_root.join(&version_prefix)?;
+
+    Ok(fs_client
+        .list_from(&start_from)?
+        .map(|meta| ParsedLogPath::try_from(meta?))
+        // TODO this filters out .crc files etc which start with "." - how do we want to use these kind of files?
+        .filter_map_ok(identity)
+        .take_while(move |path_res| match path_res {
+            Ok(path) => end_version.is_none_or(|end_version| path.version <= end_version),
+            Err(_) => true,
+        }))
+}
+/// List all commit and checkpoint files with versions above the provided `start_version` (inclusive).
+/// If successful, this returns a tuple `(sorted_commit_files_paths, checkpoint_parts)` of type
+/// `(Vec<ParsedLogPath>, Vec<ParsedLogPath>)`. The commit files are guaranteed to be sorted in
+/// ascending order by version. The elements of `checkpoint_parts` are all the parts of the same
+/// checkpoint. Checkpoint parts share the same version.
+fn list_log_files_with_version(
+    fs_client: &dyn FileSystemClient,
+    log_root: &Url,
+    start_version: Option<Version>,
+    end_version: Option<Version>,
+) -> DeltaResult<(Vec<ParsedLogPath>, Vec<ParsedLogPath>)> {
+    let mut max_checkpoint_version = start_version;
+    let mut checkpoint_parts = vec![];
+    // We expect 10 commit files per checkpoint, so start with that size. We could adjust this based
+    // on config at some point
+    let mut commit_files = Vec::with_capacity(10);
```

Comment:
tiny readability nit:
```suggestion
    // We expect 10 commit files per checkpoint, so start with that size. We could adjust this based
    // on config at some point
    let mut commit_files = Vec::with_capacity(10);
    let mut checkpoint_parts = vec![];
    let mut max_checkpoint_version = start_version;
```